File: leetcode/201-300/248-strobogrammatic-number-iii/bench/strobogrammatic_count.py
"""Benchmark workload for LeetCode #248 — Strobogrammatic Number III (Python; scale lane)."""

PAIR_A = ["0", "1", "6", "8", "9"]
PAIR_B = ["0", "1", "9", "8", "6"]


def pow5(e):
    a = 1
    for _ in range(e):
        a *= 5
    return a


def count_of_length(l):
    if l <= 0:
        return 0
    if l == 1:
        return 3
    t = 4 * pow5(l // 2 - 1)
    if l % 2 == 1:
        t *= 3
    return t


def build(k, n):
    if k == 0:
        return [""]
    if k == 1:
        return ["0", "1", "8"]
    out = []
    for s in build(k - 2, n):
        for p in range(5):
            if PAIR_A[p] == "0" and k == n:
                continue
            out.append(PAIR_A[p] + s + PAIR_B[p])
    return out


def cmp_digits(a, b):
    if len(a) != len(b):
        return -1 if len(a) < len(b) else 1
    for i in range(len(a)):
        if a[i] != b[i]:
            return -1 if a[i] < b[i] else 1
    return 0


def count_bounded(l, low, high, use_lo, use_hi):
    n = 0
    for s in build(l, l):
        keep = True
        if use_lo and cmp_digits(s, low) < 0:
            keep = False
        if use_hi and cmp_digits(s, high) > 0:
            keep = False
        if keep:
            n += 1
    return n


def count_in_range(low, high):
    ll, hl = len(low), len(high)
    if ll > hl:
        return 0
    if ll == hl:
        if cmp_digits(low, high) > 0:
            return 0
        return count_bounded(ll, low, high, True, True)
    total = count_bounded(ll, low, high, True, False)
    total += count_bounded(hl, low, high, False, True)
    for l in range(ll + 1, hl):
        total += count_of_length(l)
    return total


def main():
    queries = 1000
    state = 248248
    sink = 0
    for _ in range(queries):
        state = (state * 1103515245 + 12345) & 2147483647
        da = (state // 65536) % 8 + 1
        state = (state * 1103515245 + 12345) & 2147483647
        db = (state // 65536) % 8 + 1
        state = (state * 1103515245 + 12345) & 2147483647
        ra = (state // 65536) % 9000 + 1
        state = (state * 1103515245 + 12345) & 2147483647
        rb = (state // 65536) % 9000 + 1

        a = ra
        for i in range(1, da):
            a = a * 10 % 1000000000000000 + (i % 10)
        b = rb
        for j in range(1, db):
            b = b * 10 % 1000000000000000 + (j % 10)
        if a > b:
            a, b = b, a

        sink = (sink + count_in_range(str(a), str(b))) % 1000000007
    print(sink)


main()

File: leetcode/201-300/248-strobogrammatic-number-iii/bench/strobogrammatic_count.rs
// Benchmark workload for LeetCode #248 — Strobogrammatic Number III (Rust mirror).
// Mirrors strobogrammatic_count.kara algorithm-for-algorithm: closed form for
// interior lengths, boundary lengths enumerated and filtered.

const PAIR_A: [&str; 5] = ["0", "1", "6", "8", "9"];
const PAIR_B: [&str; 5] = ["0", "1", "9", "8", "6"];

fn pow5(e: i64) -> i64 {
    let mut acc = 1i64;
    for _ in 0..e {
        acc *= 5;
    }
    acc
}

fn count_of_length(len: i64) -> i64 {
    if len <= 0 {
        return 0;
    }
    if len == 1 {
        return 3;
    }
    let mut total = 4 * pow5(len / 2 - 1);
    if len % 2 == 1 {
        total *= 3;
    }
    total
}

fn build(k: i64, n: i64) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    if k == 0 {
        out.push(String::new());
        return out;
    }
    if k == 1 {
        out.push("0".to_string());
        out.push("1".to_string());
        out.push("8".to_string());
        return out;
    }
    let inner = build(k - 2, n);
    for s in &inner {
        for p in 0..5 {
            if PAIR_A[p] == "0" && k == n {
                continue;
            }
            let mut t = String::from(PAIR_A[p]);
            t.push_str(s);
            t.push_str(PAIR_B[p]);
            out.push(t);
        }
    }
    out
}

fn cmp_digits(a: &str, b: &str) -> i64 {
    let (xs, ys) = (a.as_bytes(), b.as_bytes());
    if xs.len() != ys.len() {
        return if xs.len() < ys.len() { -1 } else { 1 };
    }
    for i in 0..xs.len() {
        if xs[i] != ys[i] {
            return if xs[i] < ys[i] { -1 } else { 1 };
        }
    }
    0
}

fn count_bounded(len: i64, low: &str, high: &str, use_lo: bool, use_hi: bool) -> i64 {
    let cands = build(len, len);
    let mut n = 0i64;
    for s in &cands {
        let mut keep = true;
        if use_lo && cmp_digits(s, low) < 0 {
            keep = false;
        }
        if use_hi && cmp_digits(s, high) > 0 {
            keep = false;
        }
        if keep {
            n += 1;
        }
    }
    n
}

fn count_in_range(low: &str, high: &str) -> i64 {
    let lo_len = low.as_bytes().len() as i64;
    let hi_len = high.as_bytes().len() as i64;
    if lo_len > hi_len {
        return 0;
    }
    if lo_len == hi_len {
        if cmp_digits(low, high) > 0 {
            return 0;
        }
        return count_bounded(lo_len, low, high, true, true);
    }
    let mut total = count_bounded(lo_len, low, high, true, false);
    total += count_bounded(hi_len, low, high, false, true);
    let mut len = lo_len + 1;
    while len < hi_len {
        total += count_of_length(len);
        len += 1;
    }
    total
}

fn digits_of(v: i64) -> String {
    if v == 0 {
        return "0".to_string();
    }
    let mut rev: Vec<i64> = Vec::new();
    let mut x = v;
    while x > 0 {
        rev.push(x % 10);
        x /= 10;
    }
    let mut s = String::new();
    for i in (0..rev.len()).rev() {
        s.push_str(&format!("{}", rev[i]));
    }
    s
}

fn main() {
    let queries: i64 = 1000;
    let mut state: i64 = 248248;
    let mut sink: i64 = 0;

    for _ in 0..queries {
        state = (state * 1103515245 + 12345) & 2147483647;
        let da = (state / 65536) % 8 + 1;
        state = (state * 1103515245 + 12345) & 2147483647;
        let db = (state / 65536) % 8 + 1;

        state = (state * 1103515245 + 12345) & 2147483647;
        let ra = (state / 65536) % 9000 + 1;
        state = (state * 1103515245 + 12345) & 2147483647;
        let rb = (state / 65536) % 9000 + 1;

        let mut a = ra;
        let mut i = 1i64;
        while i < da {
            a = a * 10 % 1000000000000000 + (i % 10);
            i += 1;
        }
        let mut b = rb;
        let mut j = 1i64;
        while j < db {
            b = b * 10 % 1000000000000000 + (j % 10);
            j += 1;
        }
        if a > b {
            std::mem::swap(&mut a, &mut b);
        }

        let n = count_in_range(&digits_of(a), &digits_of(b));
        sink = (sink + n) % 1000000007;
    }
    println!("{}", sink);
}
